src/docfx/build/redirection/RedirectionProvider.cs

```diff
@@ -45,8 +45,14 @@ public string GetRedirectUrl(FilePath file)
 
         public FilePath GetOriginalFile(FilePath file)
         {
+            var redirectionChain = new HashSet<FilePath>();
             while (_renameHistory.TryGetValue(file, out var renamedFrom))
             {
+                if (redirectionChain.Contains(file))
```

Comment:
`HashSet.Add` method returns `false` if an item already exists in the collection, so can simplify to:
```suggestion
                if (!redirectionChain.Add(file))
                {
                    return file;
                }
```